ecleankernel/file.py

```diff
@@ -101,18 +138,37 @@ def read_internal_version(self) -> str:
             raise UnrecognizedKernelError(
                 f'Kernel file {self.path} terminates before bzImage '
                 f'header')
-        if buf[2:6] != b'HdrS':
-            raise UnrecognizedKernelError(
-                f'Unmatched magic for kernel file {self.path} '
-                f'({repr(buf[2:6])} != b"HdrS")')
-        offset = struct.unpack_from('H', buf, 0x0e)[0]
-        f.seek(offset - 0x10, 1)
-        buf = f.read(0x100)  # XXX
-        if not buf:
+        if buf[2:6] == b'HdrS':
+            offset = struct.unpack_from('H', buf, 0x0e)[0]
+            f.seek(offset - 0x10, 1)
+            buf = f.read(0x100)  # XXX
+            if not buf:
+                raise UnrecognizedKernelError(
+                    f'Kernel file {self.path} terminates before expected '
+                    f'version string position ({offset + 0x200})')
+            ret = buf.decode(errors='ignore').split(' ', 1)
+        else:
+            # If it's not a bzImage it must be a raw binary,
+            # check if it's compressed first
+            b = self.decompress_raw()
+            # unlike with bzImage, the raw kernel binary has no header
+            # that includes the version, so we parse the version message
+            # that appears on boot
+            ver_start = 'Linux version '
+            pos = b.find(ver_start.encode())
+            if pos == -1:
+                raise UnrecognizedKernelError(
+                    f'Kernel file {self.path} does not appear '
+                    f'to have a version string, '
+                    f'or the compression format was not recognized')
+            pos += len(ver_start)
+            sbuf = b[pos:pos + 0x100]
+            ret = sbuf.decode(errors='ignore').split(' ', 1)
+        if len(ret) == 1:
             raise UnrecognizedKernelError(
-                f'Kernel file {self.path} terminates before expected '
-                f'version string position ({offset + 0x200})')
-        return buf.split(b' ', 1)[0].decode()
+                f'Kernel file {self.path} terminates '
+                f'before end of version string')
+        return (ret[0])
```

Comment:
...and then:

```suggestion
        return ret[0].decode('ASCII')
```

I suppose we shouldn't ignore errors at this point as kernel version should be plain ASCII, right?

Reply:
Very good change - indeed no need to do errors=ignore - I omitted specifying the encoding now, couldn't measure a speed difference for the just 0x100 bytes we decode now